src/freenet/client/async/HealingDecisionSupplier.java

```diff
@@ -0,0 +1,74 @@
+package freenet.client.async;
+
+import freenet.node.Location;
+import freenet.node.Node;
+import freenet.node.NodeStarter;
+
+    /**
+     * Specialize Healing to the fraction of the keyspace in which we would receive the inserts
+     * if we were one of 5 long distance nodes of an actual inserter.
+     *
+     * If an opennet node is connected to an attacker, healing traffic could be mistaken for an insert.
+     * Since opennet cannot be fully secured, this should be avoided.
+     * As a solution, we specialize healing inserts to the inserts we would send if we were one of 5
+     * long distance connections for a node in another part of the keyspace.
+     *
+     * As a welcome side effect, specialized healing inserts should take one hop less to reach the
+     * correct node from which loop detection will stop the insert long before HTL reaches zero.
+     */
+public class HealingDecisionSupplier {
+  private final Node node;
+
+  public HealingDecisionSupplier(Node node) {
+    this.node = node;
+  }
+
+  public boolean shouldHeal(double keyLocation) {
+    if (!node.isOpennetEnabled()) {
+      // darknet is safer against sybil attack, so we can heal fully
+      return true;
+    }
+    double randomBetweenZeroAndOne = NodeStarter.getGlobalSecureRandom().nextDouble();
+    return shouldHealBlock(node.getLocation(), keyLocation, randomBetweenZeroAndOne);
+  }
+
+  /**
+   * Specialize healing: we want healing traffic to look like regular forwarding.
+   *
+   * Far away blocks would be unlikely to reach our node as request,
+   * so we reduce healing there: only accept 10% of those.
+   *
+   * When a key is close to our location, we use a continuous function depending on the distance
+   * to choose a probability. The closer to our node, the higher the probability of healing.
+   * As a side effect this reduces the hops healing inserts take, reducing the overall load
+   * on the network.
+   *
+   * The continuous function is gauged to accept 50% of the keys close to our node: a peak at our
+   * own location for which the area below the curve between 0 and 0.1 sums up to 0.5.
+   *
+   * Close keys are those in our 20% of the keyspace: the ones that would reach us if we were one of
+   * 5 long distance peers of a peer node. These are the keys for which we are most likely to be
+   * the best next hop when seen from the originator.
+   */
+  static boolean shouldHealBlock(
```

Comment:
That worked out pretty well — thank you!

I provide a package-private constructor now to pass in a random value provider so this can be tested deterministically, and it now looks pretty clean to me.